src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1869,24 +1891,34 @@ bool isRefEscape
                 isRefEscape,
                 ignoreArglistRefKinds: true, // https://github.com/dotnet/roslyn/issues/63325: for compatibility with C#10 implementation.
                 argsAndParamsAll);
+
+            var returnsWritableRefToRefStruct = ReturnsWritableRefToRefStruct(symbol);
             foreach (var argAndParam in argsAndParamsAll)
             {
-                var argument = argAndParam.Argument;
-                bool valid = argAndParam.IsRefEscape ?
-                    CheckRefEscape(argument.Syntax, argument, escapeFrom, escapeTo, false, diagnostics) :
-                    CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, false, diagnostics);
-
-                if (!valid)
+                // SPEC:
+                // If `M()` does return ref-to-ref-struct, the *safe-to-escape* is the same as the *safe-to-escape* of all arguments which are ref-to-ref-struct.
+                // If `M()` does return ref-to-ref-struct, the *ref-safe-to-escape* is the narrowest *ref-safe-to-escape* contributed by all arguments which are ref-to-ref-struct.
+                //
+                if (!returnsWritableRefToRefStruct
+                    || (argAndParam.Parameter?.RefKind.IsWritableReference() == true && argAndParam.Parameter.Type.IsRefLikeType && argAndParam.IsRefEscape == isRefEscape))
```

Comment:
>argAndParam.Parameter [](http://example.com/codeflow?start=24&length=21)

Consider extracting a `param` local, or using `foreach` deconstruction above to match `GetInvocationEscapeWithUpdatedRules()`: `foreach (var (param, argument, _, isArgumentRefEscape) in argsAndParamsAll)`